File: src/trajectory_reader.cpp
#include "amr_trajectory_tools/trajectory_reader.hpp"

/*
# Author:    < Alen Gigi >
# Filename:  < trajectory_reader.cpp >
# Functions: < Publish the saved trajectory data for visualization >
*/

// Initialize the node
TrajectoryReader::TrajectoryReader() : Node("trajectory_reader_node") {
    // Declare parameters with default values
    declare_parameter("marker_topic", "/read_trajectory_markers");
    declare_parameter("frame", "odom");

    // Get parameter values
    frame_ = get_parameter("frame").as_string();
    marker_topic_ = get_parameter("marker_topic").as_string();

    // Publisher for trajectory markers
    marker_pub_ = create_publisher<visualization_msgs::msg::MarkerArray>(marker_topic_, 10);

    // Service to trigger saved trajectory publishing
    load_service_ = create_service<amr_trajectory_tools::srv::ReadTrajectoryMsg>(
        "read_trajectory", std::bind(&TrajectoryReader::read_trajectory, this, std::placeholders::_1, std::placeholders::_2));

    // Some loggings
    RCLCPP_INFO(get_logger(), "Trajectory Reader Node Initialized.");
    RCLCPP_INFO(get_logger(), "To visualize a saved trajectory, call the service /read_trajectory with the path of the JSON file.");
}

void TrajectoryReader::read_trajectory(
    const std::shared_ptr<amr_trajectory_tools::srv::ReadTrajectoryMsg::Request> request,
    std::shared_ptr<amr_trajectory_tools::srv::ReadTrajectoryMsg::Response> response) {
    try {
        // Get filename from request
        std::ifstream file(request->filename);
        Json::Value trajectory_data;
        file >> trajectory_data;

        // Handling case where the data is empty
        if (trajectory_data.empty()) {
            RCLCPP_WARN(get_logger(), "Trajectory file is empty.");
            response->success = false;
            response->message = "File is empty.";
            return;
        }

        // Function call to publish the trajectory data
        publish_markers(trajectory_data);
        RCLCPP_INFO(get_logger(), "Trajectory loaded from %s and published.", request->filename.c_str());
        response->success = true;
        response->message = "Successfully loaded trajectory.";
    } catch (const std::exception &e) {
        RCLCPP_ERROR(get_logger(), "Failed to load trajectory: %s", e.what());
        response->success = false;
        response->message = "Error loading trajectory.";
    }
}

void TrajectoryReader::publish_markers(const Json::Value &trajectory_data) {
    // Function to publish the trajectory as a MarkerArray
    visualization_msgs::msg::MarkerArray marker_array;

    for (Json::ArrayIndex i = 0; i < trajectory_data.size(); ++i) {
        visualization_msgs::msg::Marker marker;
        marker.header.frame_id = frame_; // Frame to publish the data
        marker.header.stamp = now();
        marker.ns = "loaded_trajectory";
        marker.id = i;
        marker.type = visualization_msgs::msg::Marker::SPHERE; // Type of the marker
        marker.action = visualization_msgs::msg::Marker::ADD;

        // Extract x, y from the saved trajectory
        marker.pose.position.x = trajectory_data[i]["x"].asFloat();
        marker.pose.position.y = trajectory_data[i]["y"].asFloat();

        // Convert yaw data back to quaternion
        float yaw = trajectory_data[i]["yaw"].asFloat();
        marker.pose.orientation.z = std::sin(yaw / 2);
        marker.pose.orientation.w = std::cos(yaw / 2);

        // Dimensions of the marker x, y, z
        marker.scale.x = marker.scale.y = marker.scale.z = 0.08;

        // Color of the marker (green)
        marker.color.a = 1.0;
        marker.color.r = 0.0;
        marker.color.g = 1.0;
        marker.color.b = 0.0;

        marker_array.markers.push_back(marker);
    }

    marker_pub_->publish(marker_array); // Publish the marker data
    RCLCPP_INFO(get_logger(), "Trajectory markers published for visualization.");
}

int main(int argc, char *argv[]) {
    // Initialize rclcpp
    rclcpp::init(argc, argv);
    // Create the node instance
    auto node = std::make_shared<TrajectoryReader>();
    // Spin the node once
    rclcpp::spin(node);
    // Stop the node
    rclcpp::shutdown();
    return 0;
}

File: scripts/trajectory_reader.py
#!/usr/bin/env python3

'''
# Author:    < Alen Gigi >
# Filename:  < trajectory_reader.py >
# Functions: < Publish the saved trajectory data for visualization >
'''

#necesary imports
import rclpy
import json
import math
from rclpy.node import Node
from visualization_msgs.msg import Marker, MarkerArray
from amr_trajectory_tools.srv import ReadTrajectoryMsg  # Import the custom service
import math

#trajectory reader class
class TrajectoryReader(Node):
    def __init__(self):

        # Initialize the node
        super().__init__('trajectory_reader_node')

        # Declare parameters with default values
        self.declare_parameter(
            'marker_topic', '/read_trajectory_markers'
        )
        self.declare_parameter(
            'frame', 'odom'
        )

        # Get parameter values
        self.frame = self.get_parameter('frame').value
        self.marker_topic = self.get_parameter('marker_topic').value

        # Publisher for trajectory markers
        self.marker_pub = self.create_publisher(MarkerArray, self.marker_topic, 10)

        # Service to trigger saved trajectory publishing
        self.load_service = self.create_service(ReadTrajectoryMsg , 'read_trajectory', self.read_trajectory)

        #some loggings..
        self.get_logger().info("Trajectory Reader Node Initialized.")
        self.get_logger().info("To Visualize a saved Trajectory, call the service /read_trajectory with the path of the Json file.")

    def read_trajectory(self, request, response):
        """Service callback to read given trajectory file and publish markers."""
        try:
            # Get filename from request
            file_path = request.filename
            
            #open the file in readmode
            with open(file_path, 'r') as file:
                trajectory_data = json.load(file)
            
            #handling case where the data is empty
            if not trajectory_data:
                self.get_logger().warn("Trajectory file is empty.")
                response.success = False
                response.message = "File is empty."
                return response
            
            #function call to publish the traectory data
            self.publish_markers(trajectory_data)
            self.get_logger().info(f"Trajectory loaded from {file_path} and published.")
            response.success = True
            response.message = f"Successfully loaded trajectory from {file_path}"
        except Exception as e:
            self.get_logger().error(f"Failed to load trajectory: {str(e)}")
            response.success = False
            response.message = f"Error: {str(e)}"

        return response

    def publish_markers(self, trajectory_data):
        """Fuction to publish the trajectory as a MarkerArray."""
        marker_array = MarkerArray()

        for i, point in enumerate(trajectory_data):
            marker = Marker()
            marker.header.frame_id = self.frame  #frame to publish the data
            marker.header.stamp = self.get_clock().now().to_msg()
            marker.ns = 'loaded_trajectory'
            marker.id = i
            marker.type = Marker.SPHERE #type of the marker
            marker.action = Marker.ADD

            # Extract x, y from the saved trajectory
            marker.pose.position.x = point['x']
            marker.pose.position.y = point['y']

            # Convert yaw data back to quaternion
            qz = math.sin(point['yaw'] / 2)
            qw = math.cos(point['yaw'] / 2)
            marker.pose.orientation.z = qz
            marker.pose.orientation.w = qw
            
            #dimensions of the marker x,y,z
            marker.scale.x = 0.08
            marker.scale.y = 0.08
            marker.scale.z = 0.08
            
            #color of the marker rgb and alpha (green)
            marker.color.a = 1.0
            marker.color.r = 0.0
            marker.color.g = 1.0
            marker.color.b = 0.0
            marker.lifetime.sec = 0  # Keep markers permanently

            marker_array.markers.append(marker)

        self.marker_pub.publish(marker_array) #publish the marker data
        self.get_logger().info("Trajectory markers published for visualization.")


def main(args=None):
    #initialize rclpy
    rclpy.init(args=args)
    #creating the node instance
    node = TrajectoryReader()
    #spin the node once
    rclpy.spin(node)
    #stop the node
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
